### PR TITLE
New field "type" in json generated by LwM2mIdentitySerDes and LwM2mPeerSerDes.

### PR DESCRIPTION
**Pull request by JaroslawLegierski**
Implementation of field "type" in json generated by LwM2mIdentitySerDes and LwM2mPeerSerDes.

### CODE REVIEW

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mIdentitySerDes.java [hunk 1]
```diff
@@ -63,34 +72,40 @@ public JsonNode serialize(LwM2mIdentity identity) {
...
         }
 
-        JsonNode jpsk = jObj.get(KEY_ID);
-        if (jpsk != null) {
-            return new PskIdentity(jpsk.asText());
+        if ((jObj.get(KEY_LWM2MIDENTITY_TYPE).asText()).equals(LWM2MIDENTITY_TYPE_PSK)) {
+            JsonNode jpsk = jObj.get(KEY_ID);
+            if (jpsk != null) {
+                return new PskIdentity(jpsk.asText());
+            }
```
Comment (sbernard31) [inline leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mIdentitySerDes.java:87]: same here

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mIdentitySerDes.java [hunk 2]
```diff
@@ -63,34 +72,40 @@ public JsonNode serialize(LwM2mIdentity identity) {
...
+                }
             }
         }
 
-        JsonNode jcn = jObj.get(KEY_CN);
-        if (jcn != null) {
-            return new X509Identity(jcn.asText());
+        if ((jObj.get(KEY_LWM2MIDENTITY_TYPE).asText()).equals(LWM2MIDENTITY_TYPE_X509)) {
+            JsonNode jcn = jObj.get(KEY_CN);
+            if (jcn != null) {
```
Comment (sbernard31) [inline leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mIdentitySerDes.java:106]: same here

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mPeerSerDes.java [hunk 3]
```diff
@@ -28,8 +28,8 @@ public class LwM2mPeerSerDes {
 
     // LwM2mPeer keys
     protected static final String KEY_IDENTITY = "identity";
-    // TODO should we add a type field here ?
-
+    protected static final String KEY_LWM2MPEERTYPE = "type";
+    protected static final String DEFAULT_LWM2MPEERTYPE = "ippeer";
```
Comment (sbernard31) [inline leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mPeerSerDes.java:32]: I'm not sure about the constant name ? why not  : 
 - `KEY_LWM2MPEER_TYPE`
 - `LWM2MPEER_TYPE_IP` ? 
 
(Same logic you used at LwM2mIdentitySerDes)

Maybe the value could also be just `ip` ?

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mIdentitySerDes.java [hunk 4]
```diff
@@ -40,21 +40,30 @@ public class LwM2mIdentitySerDes {
     protected static final String KEY_ID = "pskid";
     protected static final String KEY_CN = "cn";
     protected static final String KEY_RPK = "rpk";
-    // TODO should we add a type field here ?
+    // identity field key and types
+    protected static final String KEY_LWM2MIDENTITY_TYPE = "type";
+    protected static final String LWM2MIDENTITY_TYPE_UNSECURE = "unsecure";
+    protected static final String LWM2MIDENTITY_TYPE_PSK = "psk";
+    protected static final String LWM2MIDENTITY_TYPE_X509 = "cn";
```
[the remark this answers is not in the document]
Comment (JaroslawLegierski) [reply leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mIdentitySerDes.java:47]: OK no problem we can use "x509"

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mIdentitySerDes.java [hunk 5]
```diff
@@ -40,21 +40,30 @@ public class LwM2mIdentitySerDes {
...
+    protected static final String LWM2MIDENTITY_TYPE_RPK = "rpk";
+    protected static final String LWM2MIDENTITY_TYPE_RID = "rid";
```
[the remark this answers is not in the document]
Comment (JaroslawLegierski) [reply leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mIdentitySerDes.java:49]: OK - I removed the unused "rid" key

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mIdentitySerDes.java [hunk 6]
```diff
@@ -63,34 +72,40 @@ public JsonNode serialize(LwM2mIdentity identity) {
...
-        JsonNode jaddress = jObj.get(KEY_ADDRESS);
-        JsonNode jport = jObj.get(KEY_PORT);
-        if (jaddress != null && jport != null) {
-            return new SocketIdentity(new InetSocketAddress(jaddress.asText(), jport.asInt()));
+        if ((jObj.get(KEY_LWM2MIDENTITY_TYPE).asText()).equals(LWM2MIDENTITY_TYPE_UNSECURE)) {
+            JsonNode jaddress = jObj.get(KEY_ADDRESS);
+            JsonNode jport = jObj.get(KEY_PORT);
+            if (jaddress != null && jport != null) {
+                return new SocketIdentity(new InetSocketAddress(jaddress.asText(), jport.asInt()));
+            }
```
[the remark this answers is not in the document]
Comment (JaroslawLegierski) [reply leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mIdentitySerDes.java:80]: OK already corrected in newest commit

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mIdentitySerDes.java (hunk 1 above)
Comment (JaroslawLegierski) [reply leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mIdentitySerDes.java:87]: corrected

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mIdentitySerDes.java [hunk 7]
```diff
@@ -63,34 +72,40 @@ public JsonNode serialize(LwM2mIdentity identity) {
...
-            try {
-                byte[] rpk = Hex.decodeHex(jrpk.asText().toCharArray());
-                X509EncodedKeySpec spec = new X509EncodedKeySpec(rpk);
-                PublicKey publicKey = KeyFactory.getInstance("EC").generatePublic(spec);
-                return new RpkIdentity(publicKey);
-            } catch (InvalidKeySpecException | NoSuchAlgorithmException e) {
-                throw new IllegalStateException("Invalid security info content", e);
+        if ((jObj.get(KEY_LWM2MIDENTITY_TYPE).asText()).equals(LWM2MIDENTITY_TYPE_RPK)) {
+            JsonNode jrpk = jObj.get(KEY_RPK);
+            if (jrpk != null) {
```
[the remark this answers is not in the document]
Comment (JaroslawLegierski) [reply leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mIdentitySerDes.java:92]: corrected

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mIdentitySerDes.java (hunk 2 above)
Comment (JaroslawLegierski) [reply leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mIdentitySerDes.java:106]: corrected

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mPeerSerDes.java (hunk 3 above)
Comment (JaroslawLegierski) [reply leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mPeerSerDes.java:32]: > I'm not sure about the constant name ? why not :
> 
>     * `KEY_LWM2MPEER_TYPE`
> 
>     * `LWM2MPEER_TYPE_IP` ?
> 
> 
> (Same logic you used at LwM2mIdentitySerDes)

OK maybe it's better to unify it. 
 
> Maybe the value could also be just `ip` ?

"ip" also  sounds better than "ippeer"

File: leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mPeerSerDes.java [hunk 8]
```diff
@@ -42,21 +42,24 @@ public JsonNode serialize(LwM2mPeer peer) {
...
+        if ((jObj.get(KEY_LWM2MPEERTYPE).asText()).equals(DEFAULT_LWM2MPEERTYPE)) {
+            String address = jObj.get(KEY_ADDRESS).asText();
+            int port = jObj.get(KEY_PORT).asInt();
 
-        return new IpPeer(new InetSocketAddress(address, port), identitySerDes.deserialize(jObj.get("identity")));
+            return new IpPeer(new InetSocketAddress(address, port), identitySerDes.deserialize(jObj.get("identity")));
+        } else {
+            throw new IllegalStateException(String.format("Invalid object type, expected type for this object is: %s",
+                    jObj.get(KEY_LWM2MPEERTYPE)));
+        }
```
[the remark this answers is not in the document]
Comment (JaroslawLegierski) [reply leshan-server-redis/src/main/java/org/eclipse/leshan/server/redis/serialization/LwM2mPeerSerDes.java:62]: OK it sounds better

**Comment by sbernard31** [comment]
I Integrated this in `identity_refactoring` branch. (squashed in one commit 26a050a8c362e1ebb08106244cff3f2085c1758d) 

@JaroslawLegierski there is a way to launch build to test Redis stores. Personally I use : `mvn clean install javadoc:javadoc -Predis`

The `redis` profile can only be launched locally with a redis server running on localhost by default (it could be changed using `REDIS_URI` environnement variable). This is not done by jenkins OR github action because we have no redis instance available.

When I did it,  `RedisRegistrationTest` failed :x:.
With [this fix](https://github.com/eclipse/leshan/commit/26a050a8c362e1ebb08106244cff3f2085c1758d#diff-94dae8ca6c49ebf1a21ef36d9913889eb9877163001d657215c943632eb59e90R1002-R1006) that sounds OK :heavy_check_mark:.  (This fix in integrated in squashed commit)

I guess we can close this issue now.

**Pull request closed without merging**